tinygrad/runtime/ops_cuda.py

```diff
@@ -49,9 +49,16 @@ class device:
   import pycuda.autoprimaryctx # type: ignore # pylint: disable=unused-import # noqa: F401
   import pycuda.driver as cuda # type: ignore
   class CUDAAllocator(LRUAllocator):
-    def _do_alloc(self, size, dtype, device, **kwargs): return cuda.mem_alloc(size * dtype.itemsize) # type: ignore
+    def _do_alloc(self, size, dtype, device, **kwargs):
+      try: return cuda.mem_alloc(size * dtype.itemsize)
+      except cuda.MemoryError as e:
+        if len(self.aging_order[device]):
+          self.ensure_has_free_space(1.1*self._get_cur_free_spce(device), device) # increase free space by 10% and try again.
```

Comment:
Maybe move this logic into LRU?

Something like?
```python
def _alloc_buffer(self, size, dtype, device, **kwargs):
    newbuf, alloc_size = None, size*dtype.itemsize
    while newbuf is None:
        self.ensure_has_free_space(alloc_size, device)
        try: newbuf = self._do_alloc(max(1, size), dtype, device, **kwargs)
        except Exception as e:
            if not len(self.aging_order[device]): raise e
            alloc_size *= 1.1
    self.free_space[device] -= alloc_size
    self.buffer_info[newbuf] = (size, dtype, alloc_size, device)
    return newbuf
```

Reply:
Different backend will raise different exceptions, either we handle any exception, which can be unrelated to memory, or we do an 'or' of all known memory error from each backend, which will result in backend related imports in lib.py or we introduce our memory exception and wrap backend exception in it handle that in lib.py. To avoid all that I have just handled it cuda file.

Alloc is costly, finding free space on gpu is not. I am not sure if unsuccessful alloc is fail fast. Also I have observed it diverges significantly for free_space calculation.

Reply:
I don't think it belongs to cuda file, since it's related to all backends. I think it's fine to catch all exceptions.

Reply:
Done.